### PR TITLE
Shape inference for Cast

### PR DESCRIPTION
**Pull request by anderspapitto**

### CODE REVIEW

**Comment by anderspapitto** [comment]
built on top of https://github.com/onnx/onnx/pull/655

File: onnx/defs/schema.cc [hunk 1]
```diff
@@ -584,4 +584,54 @@ size_t ReplaceAll(std::string& s, const char* from, const char* to) {
   }
   return numReplaced;
 }
+
+TensorProto::DataType datatypeFromString(const std::string& type) {
+  if (type == "FLOAT") {
+    return TensorProto::FLOAT;
+  }
+  if (type == "FLOAT") {
```
Comment (houseroad) [inline onnx/defs/schema.cc]: 2 FLOAT? C++ may miss some float, so we need a second check to catch them?

Comment (anderspapitto) [reply onnx/defs/schema.cc]: float is a very important datatype and needs extra attention

File: onnx/defs/shape_inference.h [hunk 2]
```diff
@@ -32,6 +32,18 @@ inline bool getRepeatedAttribute(InferenceContext& ctx,
 
 }
 
+inline bool hasExactlyNInputTypes(InferenceContext& ctx, int n, const std::string& opname) {
+  if (ctx.getNumInputs() != n) {
+    throw std::runtime_error(opname + " has wrong number of inputs");
```
Comment (gramalingam) [inline onnx/defs/shape_inference.h]: Why not return false, instead of throwing an exception? E.g., I see this being invoked for Concat, where this may not be an error.

File: onnx/defs/tensor/defs.cc [hunk 3]
```diff
@@ -136,7 +171,70 @@ ONNX_OPERATOR_SCHEMA(Concat)
...
-        "Constrain output types to float tensors.");
+        "Constrain output types to float tensors.")
+    .ShapeInferenceFunction([](InferenceContext& ctx) {
+        if (!hasExactlyNInputTypes(ctx, 1, "Concat")) {
+          return;
+        }
+
+        propagateElemTypeFromInputToOutput(ctx, 0, 0);
+
+        if (ctx.getNumInputs() == 0) {
```
Comment (gramalingam) [inline onnx/defs/tensor/defs.cc]: I don't understand: there's a check earlier for exactly 1 input type.

File: onnx/defs/tensor/defs.cc [hunk 4]
```diff
@@ -99,7 +118,19 @@ Takes a tensor as input and outputs an 1D int64 tensor containing the shape of t
     .TypeConstraint(
         "T1",
         {"tensor(int64)"},
-        "Constrains output to int64 tensor.");
+        "Constrains output to int64 tensor.")
+    .ShapeInferenceFunction([](InferenceContext& ctx) {
+        if (!hasExactlyNInputTypes(ctx, 1, "Cast")) {
```
Comment (gramalingam) [inline onnx/defs/tensor/defs.cc]: "Cast" -> "Shape"

File: onnx/defs/shape_inference.h (hunk 2 above)
Comment (anderspapitto) [reply onnx/defs/shape_inference.h]: well, this is just a helper that throws an exception. the error is with concat - i'll remove that usage

File: onnx/defs/tensor/defs.cc (hunk 3 above)
Comment (anderspapitto) [reply onnx/defs/tensor/defs.cc]: ++

File: onnx/defs/tensor/defs.cc (hunk 4 above)
Comment (anderspapitto) [reply onnx/defs/tensor/defs.cc]: ++

File: onnx/defs/tensor/defs.cc [hunk 5]
```diff
@@ -156,7 +250,38 @@ ONNX_OPERATOR_SCHEMA(Split)
...
+          }
+          const auto& splitDim = ctx.getInputType(0)->shape().dim(axis);
+          if (!splitDim.has_dim_value()) {
+            return;
+          }
+          int splitDimValue = splitDim.dim_value();
+          int chunkSize = splitDimValue / ctx.getNumOutputs();
+          int leftOver = splitDimValue - (chunkSize * ctx.getNumOutputs());
+          for (int i = 0; i < ctx.getNumOutputs(); i++) {
+            split.push_back(i < leftOver ? chunkSize + 1 : chunkSize);
```
Comment (gramalingam) [inline onnx/defs/tensor/defs.cc]: The spec above doesn't actually say what the behavior is when leftover is non-zero. May be a good idea to extend the spec to do that (though that should probably be a separate PR since it affects the spec).

File: onnx/defs/tensor/defs.cc [hunk 6]
```diff
@@ -327,7 +452,31 @@ Example 2:
...
+        int r = ctx.getInputType(0)->shape()->dim_size();
+        int q = ctx.getInputType(1)->shape()->dim_size();
+
+        int out_rank = q + r - 1;
+
+        if (out_rank == 0) {
+          ctx.getOutputType(0)->mutable_shape();
+        }
+        for (int i = 0; i < out_rank; ++i) {
+          ctx.getOutputType(0)->mutable_shape()->add_dim();
```
Comment (gramalingam) [inline onnx/defs/tensor/defs.cc]: Is this supposed to be initialized to some value from the input dims?

File: onnx/defs/tensor/defs.cc (hunk 5 above)
Comment (anderspapitto) [reply onnx/defs/tensor/defs.cc]: ++ will make a PR for the spec

File: onnx/defs/tensor/defs.cc (hunk 6 above)
Comment (anderspapitto) [reply onnx/defs/tensor/defs.cc]: not at the moment - this is more "rank inference" than "shape inference". It can be enhanced in a future diff

Comment (gramalingam) [reply onnx/defs/tensor/defs.cc]: It would be safer to initialize the dim with the empty string param than leave it uninitialized (neither value nor param set).

Comment (anderspapitto) [reply onnx/defs/tensor/defs.cc]: I guess representing this is still a little up in air - but see the comment I just wrote at https://github.com/onnx/onnx/pull/681#issuecomment-381221996

Comment (gramalingam) [reply onnx/defs/tensor/defs.cc]: Okay. As long as the meaning/representation is standardized and well-documented, it's fine.
